validator/accounts/status.go

```diff
@@ -0,0 +1,235 @@
+
+// MaxRequestKeys specifies the max amount of public keys allowed
+// in a single grpc request, when fetching account statuses.
+const MaxRequestKeys = 2000 // XXX: This is an arbitrary number. Used to limit time complexity
+// of sorting a single batch of status requests. Do not make this number too big.
+
+// RunStatusCommand is the entry point to the `validator status` command.
+func RunStatusCommand(
+	ctx context.Context,
+	pubkeys [][]byte,
+	withCert string,
+	endpoint string,
+	maxCallRecvMsgSize int,
+	grpcRetries uint,
+	grpcHeaders []string) error {
+	dialOpts, err := constructDialOptions(maxCallRecvMsgSize, withCert, grpcHeaders, grpcRetries)
+	if err != nil {
+		return err
+	}
+	conn, err := grpc.DialContext(ctx, endpoint, dialOpts...)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to dial beacon node endpoint at %s", endpoint)
+	}
+	statuses, err := FetchAccountStatuses(
+		ctx, ethpb.NewBeaconNodeValidatorClient(conn), pubkeys)
+	if e := conn.Close(); e != nil {
+		log.WithError(e).Error("Could not close connection to beacon node")
+	}
+	if err != nil {
+		return errors.Wrap(err, "Could not fetch account statuses from the beacon node")
+	}
+	printStatuses(statuses)
+	return nil
+}
+
+func constructDialOptions(
+	maxCallRecvMsgSize int,
+	withCert string,
+	grpcHeaders []string,
+	grpcRetries uint) ([]grpc.DialOption, error) {
+	var transportSecurity grpc.DialOption
+	if withCert != "" {
+		creds, err := credentials.NewClientTLSFromFile(withCert, "")
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not get valid credentials: %v", err)
+		}
+		transportSecurity = grpc.WithTransportCredentials(creds)
+	} else {
+		transportSecurity = grpc.WithInsecure()
+		log.Warn("You are using an insecure gRPC connection! Please provide a certificate and key to use a secure connection.")
+	}
+
+	if maxCallRecvMsgSize == 0 {
+		maxCallRecvMsgSize = 10 * 5 << 20 // Default 50Mb
+	}
+
+	md := make(metadata.MD)
+	for _, hdr := range grpcHeaders {
+		if hdr != "" {
+			ss := strings.Split(hdr, "=")
+			if len(ss) != 2 {
+				log.Warnf("Incorrect gRPC header flag format. Skipping %v", hdr)
+				continue
+			}
+			md.Set(ss[0], ss[1])
+		}
+	}
+
+	return []grpc.DialOption{
+		transportSecurity,
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(maxCallRecvMsgSize),
+			grpc_retry.WithMax(grpcRetries),
+			grpc.Header(&md),
+		),
+		grpc.WithStatsHandler(&ocgrpc.ClientHandler{}),
+		grpc.WithBlock(),
+		grpc.WithTimeout(
+			10 * time.Second /* Block for 10 seconds to see if we can connect to beacon node */),
+	}, nil
+}
+
+// FetchAccountStatuses fetches validator statuses from the BeaconNodeValidatorClient
+// for each validator public key.
+func FetchAccountStatuses(
+	ctx context.Context,
+	beaconNodeRPCProvider ethpb.BeaconNodeValidatorClient,
+	pubkeys [][]byte) ([]ValidatorStatusMetadata, error) {
+	ctx, span := trace.StartSpan(ctx, "accounts.FetchAccountStatuses")
+	defer span.End()
+
```

Comment:
Please set a context timeout or deadline.

```suggestion

ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
defer cancel()

```